test/unit/org/apache/cassandra/db/tries/MemtableTrieTestBase.java

```diff
@@ -0,0 +1,630 @@
+    {
+        Iterator<E> expected = expectedIterable.iterator();
+        Iterator<E> actual = actualIterable.iterator();
+        while (actual.hasNext() && expected.hasNext())
+        {
+            Assert.assertEquals(actual.next(), expected.next());
+        }
+        if (expected.hasNext())
+            Assert.fail("Remaing values in expected, starting with " + expected.next());
+        else if (actual.hasNext())
+            Assert.fail("Remaing values in actual, starting with " + actual.next());
+    }
+
+    static ByteComparable[] generateKeys(Random rand, int count)
+    {
+        ByteComparable[] sources = new ByteComparable[count];
+        TreeSet<ByteComparable> added = new TreeSet<>((bytes1, bytes2) -> ByteComparable.compare(bytes1, bytes2, VERSION));
+        for (int i = 0; i < count; ++i)
+        {
+            sources[i] = generateKey(rand);
+            if (!added.add(sources[i]))
+                --i;
+        }
+
+        // note: not sorted!
+        return sources;
+    }
+
+    static ByteComparable generateKey(Random rand)
+    {
+        return generateKey(rand, MIN_LENGTH, MAX_LENGTH);
+    }
+
+    static ByteComparable generateKey(Random rand, int minLength, int maxLength)
+    {
+        int len = rand.nextInt(maxLength - minLength + 1) + minLength;
+        byte[] bytes = new byte[len];
+        int p = 0;
+        int length = bytes.length;
+        while (p < length)
+        {
+            int seed = rand.nextInt(KEY_CHOICE);
+            Random r2 = new Random(seed);
+            int m = r2.nextInt(5) + 2 + p;
+            if (m > length)
+                m = length;
+            while (p < m)
+                bytes[p++] = (byte) r2.nextInt(256);
+        }
+        return ByteComparable.fixedLength(bytes);
+    }
+
+    static String asString(ByteComparable bc)
+    {
+        return bc != null ? bc.byteComparableAsString(VERSION) : "null";
+    }
+
+    <T, M> void putSimpleResolve(MemtableTrie<T> trie,
+                                 ByteComparable key,
+                                 T value,
+                                 Trie.MergeResolver<T> resolver)
+    {
+        putSimpleResolve(trie, key, value, resolver, usePut());
+    }
+
+    static <T, M> void putSimpleResolve(MemtableTrie<T> trie,
+                                        ByteComparable key,
+                                        T value,
+                                        Trie.MergeResolver<T> resolver,
+                                        boolean usePut)
+    {
+        try
+        {
+            trie.putSingleton(key,
+                              value,
+                              (existing, update) -> existing != null ? resolver.resolve(existing, update) : update,
+                              usePut);
+        }
+        catch (MemtableTrie.SpaceExhaustedException e)
+        {
+            throw Throwables.propagate(e);
```

Comment:
`Throwables.propagate` is deprecated, info [here](https://github.com/google/guava/wiki/Why-we-deprecated-Throwables.propagate). We should probably just do `throw new RuntimeException(e)`.

Reply:
Maybe make SpaceExhaustedException extend RuntimeException instead?

It would be nice to move away from opaque wrappers

Reply:
I would agree in principle. But in this case there's a non-standard failure mechanism and method of dealing with it, and there is value in forcing that fact on the library's users.

Expanded the `SpaceExhaustedException` JavaDoc to clarify.

Reply:
So... should we have a special wrapper here so that we can catch and handle it in a caller?

Reply:
This is a test, and it doesn't really matter if we wrap it, call `Assert.fail` or just declare it in the calling methods.

The non-test use [in `TrieMemtable`](https://github.com/apache/cassandra/pull/1723/files#diff-ab0610cf35d7c6a8a9cca11551d8212f5936f360ebf0c2deb5f18c06e03a9f0bR191), on the other hand, converts it to `IllegalStateException` as it is completely unexpected to reach this error when `reachedAllocationThreshold` is checked and reacted to.

Reply:
My bad, I didn’t notice it was a test. Sorry for the noise!